core/src/service/api-creator.ts

```diff
@@ -0,0 +1,230 @@
+    const toPathRepeated = compile(path);
+    return toPathRepeated(params);
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.error('path:', path, 'Error parsing url parameters');
+    throw error;
+  }
+};
+
+/**
+ * Use path to match the incoming parameters, extract query and params
+ * @param path Paths that may contain parameters, such as /fdp/:id/detail, path can not include `?`
+ * @param params The incoming parameters may be query or params
+ */
+const extractPathParams = (path: string, params?: Record<string, any>) => {
+  const keys: Key[] = [];
+  pathToRegexp(path, keys);
+  const pathParams = {} as Obj<string>;
+  const bodyOrQuery = { ...params };
+  if (keys.length > 0) {
+    keys.forEach(({ name }) => {
+      pathParams[name] = bodyOrQuery[name];
+      delete bodyOrQuery[name];
+    });
+  }
+  return {
+    pathParams,
+    bodyOrQuery,
+  };
+};
+
+interface $options {
+  isDownload?: boolean; // whether its download api
+  uploadFileKey?: string; // upload formData attribute
+  autoMessage?: boolean; // eject message automatically
+  successMsg?: string; // eject message when success to override default message
+  errorMsg?: string; // eject message when failed to override default message
+}
+type $headers = Record<string, string>;
+type $body = Record<string, any>;
+interface CallParams {
+  $options?: $options;
+  $headers?: $headers;
+  $body?: $body;
+}
+
+type Merge<A, B> = { [K in keyof A]: K extends keyof B ? B[K] : A[K] } & B extends infer O
+  ? { [K in keyof O]: O[K] }
+  : never;
+
+/**
+ * generate api request function by config
+ * @param apiConfig
+ * @returns callable api function
+ */
+export const genRequest = function <T extends FN>(apiConfig: APIConfig) {
+  const { api, headers } = apiConfig;
+  let [method, path] = api.split('@');
+  if (!path) {
+    path = method;
+    method = 'get';
+  }
+
+  return (params: CallParams & Merge<Parameters<T>[0], {}>) => {
+    const { $options, $headers, $body, ...rest } = params;
+    const { bodyOrQuery, pathParams } = extractPathParams(path, rest);
+    const { isDownload, uploadFileKey } = $options || {};
+    if ('pageNo' in bodyOrQuery && !('pageSize' in bodyOrQuery)) {
+      bodyOrQuery.pageSize = DEFAULT_PAGESIZE;
+    }
+    let bodyData;
+    if (['post', 'put'].includes(method)) {
+      bodyData = isEmpty(bodyOrQuery) ? undefined : uploadFileKey ? bodyOrQuery[uploadFileKey] : bodyOrQuery;
+    } else if (method === 'delete') {
+      bodyData = $body;
+    }
+    return axios({
+      method: method as any,
+      url: generatePath(path, pathParams),
+      headers: !isEmpty(headers) ? headers : $headers,
+      params: bodyOrQuery,
+      paramsSerializer: (p: Record<string, string>) => qs.stringify(p),
+      responseType: isDownload ? 'blob' : 'json',
+      data: bodyData,
+    }).then((res) => res.data) as unknown as Promise<BODY<ReturnType<T>>>;
+  };
+};
+
+export const apiDataStore = createStore({
+  name: 'api_data',
```

Comment:
do we have name standard for names like store name。 `_` looks a little wried

Reply:
Nope, have changed to apiData